src/signature.c

```diff
@@ -82,20 +100,38 @@ bool initialize_signature(void)
 
 	ret = validate_certificate();
 	if (ret) {
-		printf("Failed to verify certificate: %s\n", X509_verify_cert_error_string(ret));
 		if (ret == X509_V_ERR_CERT_NOT_YET_VALID) {
-			/* If we can retrieve an approx. good system time, report out to user */
-			if (stat("/usr/lib/os-release", &statt) != -1) {
-				mod_sec = statt.st_mtim.tv_sec;
-				char timebuf[30];
-				alttime = localtime(&mod_sec);
-				strftime(timebuf, sizeof(timebuf), "%F", alttime);
-				printf("System clock should be at least %s\n", timebuf);
+			time_t currtime;
+			struct tm *timeinfo;
+
+			time (&currtime);
+			timeinfo = localtime (&currtime);
+			printf ("Warning: Current time %s is incorrect\nAttempting to fix...", asctime(timeinfo));
+
+			/* If we can retrieve an approx. good system time, use it and try again */
+			if (fix_time("/usr/bin/swupd", mod_sec, alttime) == false) {
+				goto fail;
+			}
+			ret = validate_certificate();
+
+			/* If it failed try one last time to get a good system time... */
+			if (ret == X509_V_ERR_CERT_NOT_YET_VALID) {
+				if (fix_time("/usr/lib/os-release", mod_sec, alttime) == false) {
+					goto fail;
+				}
+				ret = validate_certificate();
+				if (ret) {
+					goto fail;
+				}
+				goto success;
+			} else if (ret) {
+				goto fail;
```

Comment:
This logic is pretty twisted. Maybe break this block out into another function to avoid all the gotos?